keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,989 @@
+delete the namespace too early, leaving garbage behind in etcd. This is a
+correctness problem, the garbage will reappear if a namespace of the same name
+is recreated.
+
+This proposal addresses both problems.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+1. **Network connectivity isues between apiservers**
+    
+    Cluster admins might not read the release notes and realize they should enable network/firewall connectivity between apiservers. In this case clients will receive 503s instead of transparently being proxied. 503 is still safer than today's behavior. We will clearly document the steps needed to enable the feature and also include steps to verify that the feature is working as intended. Looking at the following exposed metrics can help wth that 
+    1. `kubernetes_uvip_proxied_request_total` to monitor the number of UVIP-proxies requests. This metric can tell us the number of requests that were successfully proxied and the ones that failed
+    2. `apiserver_request_total` to check the success/error status of the requests
+
+2. **Increase in egress bandwidth**
+    
+    Requests will consume egress bandwidth for 2 apiservers when proxied. We can cap the number if needed, but upgrades aren't that frequent and few resources are changed on releases, so these requests should not be common. We will count them with a metric.
+
+3. **Increase in request traffic directed at destination kube-apiserver**
+
+    There could be a large volume of requests for a specific resource which might result in the identified apiserver being unable to serve the proxied requests. This scenario should not occur too frequently, since resource types which have large request volume should not be added or removed during an upgrade - that would cause other problems, too.
+
+4. **Indefinite rerouting of the request**
+    
+    We should ensure at most one proxy, rather than proxying the request over and over again (if the source apiserver has an incorrect understanding of what the destination apiserver can serve). To do this, we will add a new header such as `X-Kubernetes-UVIP-Rerouted:true` to the  request once it is determined that the request cannot be served by the local apiserver and should therefore be proxied.  
+    We will remove this header after the request is received by the destination apiserver (i.e. after the proxy has happened once) at which point it will be served locally.
+
+5. **Putting IP/endpoint and trust bundle control in user hands in REST APIs**
+    
+    To prevent server-side request forgeries we will not give control over information about apiserver IP/endpoint and the trust bundle (used to authenticate server while proxying) to users via REST APIs.
```

Comment:
So is there an API server flag that builds the mapping between identity and IP?  It is still unclear to me how this will work.

Reply:
As highlighted in the [Alternatives section](https://github.com/kubernetes/enhancements/pull/4015/files/5a0ca134921229e746217e899eb5157e9dbe0a67#diff-5f6582f50a19d8d39a3a9045f6f9b6e404edd88ef829c187709437484bf308f8R948), 


we can either : 

1. store the IP info directly in the APIServerIdentity lease object (which already has the APIServerID). This approach actually doesnt mitigate the risk of putting control over this info in user hands, unless we set up a webhook to not let users meddle with the APIServerIdentity leases.
2. Or retrieve the IP from  endpoint reconciler's masterlease object (that would need to be modified to store the APIServerID as a label perhaps). This approach is the safer option to use since we would be relying on storage leases that are not exposed to the end user.

Reply:
I still do not understand what you are actually planning on doing (vs. alternatives).

Reply:
I am leaning towards using the APIServerIdentity lease object, but waiting to hear from other folks if this approach is reasonable enough for alpha.

Reply:
Updated the decision under [Identifying destination apiserver's network location](https://github.com/kubernetes/enhancements/blob/01637e17da9a0d1359a9304ad005c88234cd61ca/keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md#identifying-destination-apiservers-network-location) section. We will be writing to both

1. A new endpoint reconciler's masterleases (that will be read by apiservers to find peer ip, port info)
2. v1.Lease for users to read for visibilty/debugging